src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala

```diff
@@ -42,140 +41,115 @@ object VerilatorExecutive extends BackendExecutive {
     dut:            T,
     circuit:        Circuit,
     elaboratedAnno: AnnotationSeq,
-    targetDir:      String
-  ): Option[BackendInstance[T]] = {
-    if (elaboratedAnno.contains(CachingAnnotation)) {
```

Comment:
I honestly prefer the `if(....contains(...))` over doing a `collectFirst` with side-effects.

Reply:
Please revert back to the old structure and try to keep the diff a bit smaller.

Reply:
> I honestly prefer the `if(....contains(...))` over doing a `collectFirst` with side-effects.
Sure I can change it back, but I didn't get the part about side-effect? I think they're both the same in terms of side effects?
I had a hard time figuring out what [ this lonely `None` at #L152](https://github.com/ucb-bar/chisel-testers2/pull/309/files/557c2b27160f3fce0df2c6f60bf72cd2d6b3158e#diff-0af24921e1f6b9e31fdad8acf63750b78887fc046853645e9643fa3e3ba22597L152) (original code) was doing there.

Reply:
I think functional APIs should be used when you have a side effect free function, when you are doing things that are mostly about their side effects, I prefer to stick to the imperative features in scala.
So I would use `map` for pure functions, `foreach` when you are doing side-effects.

Reply:
The `None` is there because of the use of `return` in the code. Would probably be nice to do this more like:

```
if(....) {
} else { None }
```

Reply:
`Option.when()` looks neat in these situations, but is only available in 2.13+ (maybe also backported to `scala-collection-compat`?).

Reply:
> ```
>  if(....) {
>    ....
>   Some(...)
> } else { None }
> ```

There's a nested condition (matching of hashes) which makes it difficult to do this cleanly without a `return`.